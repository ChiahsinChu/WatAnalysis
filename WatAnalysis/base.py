from tkinter.messagebox import NO
import numpy as np
from MDAnalysis.transformations import translate, wrap
from MDAnalysis.analysis.base import AnalysisBase


NA = 6.02214076E+23
ANG_TO_CM = 1E-08


def get_com_pos(universe, elem_type):
    elem_ag = universe.select_atoms("name " + elem_type)
    p = elem_ag.positions.copy()
    ti_com_z = p.mean(axis=0)[-1]
    return ti_com_z


def center(universe, elem_type, cell=None):
    """
    TBC
    """
    # check/set dimensions of universe
    if universe.dimensions is None:
        if cell is None:
            raise AttributeError('Cell parameter required!')
        else:
            universe.dimensions = cell
    # translate and wrap
    metal_com_z = get_com_pos(universe, elem_type)
    workflow = [translate([0, 0, -metal_com_z]), wrap(universe.atoms)]
    universe.trajectory.add_transformations(*workflow)
    #return universe
    
class Density1DAnalysis(AnalysisBase):
    """
    Parameters
    ----------
    universe: AtomGroup Object
        tbc
    water_sel: List or Array
        [a, b, c, alpha, beta, gamma]
    dim: int (2)
        
    delta: float (0.1)
        tbc
    """
    def __init__(self,
                 universe,
                 water_sel='name O',
                 dim=2, 
                 delta=0.1,
                 mass=18.015):
        super().__init__(universe.trajectory)
        self._atomgroup = universe.select_atoms(water_sel)
        self._cell = universe.dimensions
        self._dim = dim
        self._delta = delta
        self._mass = mass
        self._natoms = len(self._atomgroup)
        self._mask = self._atomgroup.indices

        # check cell
        if self._cell is None:
            raise AttributeError('Cell parameters should be set.')

        #parallel value initial
        self.para = None
        self._para_region = None


    def _prepare(self):
        # cross area along the selected dimension
        dims = self._cell[: 3]
        dims = np.delete(dims, self._dim)
        self.cross_area = dims[0] * dims[1] * np.sin(self._cell[self._dim + 3] / 180 * np.pi)
        # placeholder
        self.all_coords = np.zeros((self.n_frames, self._natoms),
                                    dtype=np.float32)


    def _single_frame(self):
        ts_coord = self._ts.positions[self._mask].T[self._dim]
        np.copyto(self.all_coords[self._frame_index], ts_coord)


    def _conclude(self):
        bins = np.arange(0, self._cell[self._dim], self._delta)
        grids, density = self._get_density(self.all_coords, bins)
        self.results = {}
        self.results['grids'] = grids
        self.results['density'] = density

    def _get_density(self, coords, bins):
        density, grids = np.histogram(coords, bins=bins)
        grids = grids[:-1] + self._delta / 2
        density = (density / NA * self._mass) / (
            self.cross_area * self._delta * ANG_TO_CM**3) / self.n_frames
        return grids, density
        


    def _parallel_init(self, *args, **kwargs):
    
        start = self._para_region.start
        stop = self._para_region.stop
        step = self._para_region.step
        self._setup_frames(self._trajectory, start, stop, step)
        self._prepare()


    def run(self, start=None, stop=None, step=None, verbose=None):
        if verbose == True:
            print(" ", end='')
        super().run(start, stop, step, verbose)

        if self.para:
            block_result = self._para_block_result()
            if block_result == None:
                raise ValueError(
                    "in parallel, block result has not been defined or no data output!"
                )
            return block_result


    def _para_block_result(self, ):
        return self.results


    def _parallel_conclude(self, rawdata):
        method_attr = rawdata[-1]
        del rawdata[-1]
        self.start = method_attr[0]
        self.stop = method_attr[1]
        self.step = method_attr[2]
        self.frames = np.arange(self.start, self.stop, self.step)

        self.results = {}
        density = []
        for single_data in rawdata:
            density.append(single_data['density'])
        density = np.mean(density, axis=0)
        self.results['grids'] = single_data['grids']
        self.results['density']
        return "FINISH PARA CONCLUDE"


class InterfaceWatDensity(Density1DAnalysis):
    """
    TBC
    """
    def __init__(self, universe, water_sel='name O', dim=2, delta=0.1, **kwargs):
        super().__init__(universe, water_sel, dim, delta, mass=18.015)
        
        self._surf_ids = kwargs.get('surf_ids', None)
        if self._surf_ids is None:
            # if no surf ids provided
            slab_sel = kwargs.get('slab_sel', None)
            self._surf_natoms = kwargs.get('surf_natoms', None)
            if slab_sel is not None and self._surf_natoms is not None:
                self._slab_ids = universe.select_atoms(slab_sel).indices
            else:
                raise AttributeError('slab_sel and surf_natoms should be provided in the absence of surf_ids')


    def _prepare(self):
        super()._prepare()
        # placeholder for surface coords
        self.surf_coords = np.zeros((self.n_frames, 2), dtype=np.float32)
        #print(self.surf_ids)
    

    def _single_frame(self):
        # save surface coords
        ts_surf_lo = self._ts.positions[self.surf_ids[0]].T[self._dim]
        ts_surf_hi = self._ts.positions[self.surf_ids[1]].T[self._dim]
        ts_surf_region = self._get_surf_region(ts_surf_lo, ts_surf_hi)
        np.copyto(self.surf_coords[self._frame_index], ts_surf_region)

        # save water coords w.r.t. lower surfaces
        _ts_coord = self._ts.positions[self._mask].T[self._dim]
        ts_coord = self._ref_water(_ts_coord, ts_surf_region[0])
        np.copyto(self.all_coords[self._frame_index], ts_coord)


    def _conclude(self):
        surf_lo_ave = self.surf_coords[:, 0].mean(axis=0)
        surf_hi_ave = self.surf_coords[:, 1].mean(axis=0)
        bins = np.arange(0, (surf_hi_ave - surf_lo_ave + self._delta), self._delta)
        
        grids, density = self._get_density(self.all_coords, bins)
        self.results = {}
        self.results['grids'] = grids[:len(grids) // 2]
        self.results['density'] = (density + density[np.arange(len(density)-1, -1, -1)])[:len(grids) // 2] / 2


    def _parallel_conclude(self, rawdata):
        method_attr = rawdata[-1]
        del rawdata[-1]
        self.start = method_attr[0]
        self.stop = method_attr[1]
        self.step = method_attr[2]
        self.frames = np.arange(self.start, self.stop, self.step)

        n_grids = 0
        for single_data in rawdata:
            
            if n_grids == 0 or len(single_data['grids']) < n_grids:
                n_grids = len(single_data['grids'])
        
        _density = []
        _grids = []
        for single_data in rawdata:
            _density.append(single_data['density'][: n_grids])
            _grids.append(single_data['grids'][: n_grids])
        self.results['grids'] = np.mean(_grids, axis=0) 
        self.results['density'] = np.mean(_density, axis=0)

        return "FINISH PARA CONCLUDE"

    
    @property
    def surf_ids(self):
        """
        Get the indices of surface atoms
        """
        if self._surf_ids is None:
            from ase import Atoms
            slab = Atoms("H" + str(len(self._slab_ids)))
            slab.set_cell(self._cell)
            slab.set_pbc(True)
            pos = self._trajectory[0].positions[self._slab_ids]
            slab.set_positions(pos)
            slab.center(about=slab[0].position)
            slab.wrap()
            slab.center()

            coords = np.array([slab.positions[:, self._dim]])
            slab_ids = np.atleast_2d(self._slab_ids)
            data = np.concatenate((slab_ids, coords), axis=0)
            data = np.transpose(data)

            # sort from small coord to large coord
            data = data[data[:, 1].argsort()]
            upper_slab_ids = data[:self._surf_natoms][:, 0]
            upper_slab_ids.sort()
            lower_slab_ids = data[-self._surf_natoms:][:, 0]
            lower_slab_ids.sort()
            self._surf_ids = np.array([lower_slab_ids, upper_slab_ids], dtype=int)
        return self._surf_ids

    def _get_surf_region(self, _coord_lo, _coord_hi):
        """
        Return:
            numpy array of (coord_lo, coord_hi)
            coord_lo: average coord for lower surface
            coord_hi: average coord for upper surface
        """
        # fold all _coord_lo about _coord_lo.min()
        tmp = _coord_lo.min()
        for coord in _coord_lo:
            coord = coord + np.floor((tmp - coord) / self._cell[self._dim]) * self._cell[self._dim]
        coord_lo = np.mean(_coord_lo)
        # fold all _coord_hi about _coord_hi.max()
        tmp = _coord_hi.max()
        for coord in _coord_hi:
            coord = coord + np.floor((tmp - coord) / self._cell[self._dim]) * self._cell[self._dim]
        coord_hi = np.mean(_coord_hi)

        if coord_hi < coord_lo:
            coord_hi = coord_hi + self._cell[self._dim]

        return np.array([coord_lo, coord_hi], dtype=float)

    def _ref_water(self, water_coords, surf_lo):
        """
        water coord w.r.t. lower surfaces
        """
        for coord in water_coords:
            while coord < surf_lo:
                coord = coord + self._cell[self._dim]
        water_coords = water_coords - surf_lo
        return water_coords



        