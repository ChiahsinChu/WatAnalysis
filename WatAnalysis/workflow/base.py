# SPDX-License-Identifier: LGPL-3.0-or-later
from abc import ABC, abstractmethod
from typing import Any, Dict, List, Union

import numpy as np
from ase.cell import Cell
from MDAnalysis.analysis.base import AnalysisBase, Results
from MDAnalysis.core.universe import Universe

from WatAnalysis import utils


class PlanarInterfaceAnalysisBase(AnalysisBase):
    """
    Analysis class for studying the structure and properties of water molecules
    in a given molecular dynamics simulation.

    Parameters
    ----------
    universe : Universe
        The MDAnalysis Universe object containing the simulation data.
    surf_ids : Union[List, np.ndarray], optional
        List or array of surface atom indices of the form [surf_1, surf_2]
        where surf_1 and surf_2 are arrays containing the indices corresponding
        to the left surface and right surface, respectively.
    axis : int
        The axis perpendicular to the surfaces (default is 2).
        x=0, y=1, z=2.
    workflow : List[SingleAnalysis]
        List of analysis tasks to be performed in order.
    **kwargs : dict
        Additional keyword arguments for customization:
        - verbose : bool, optional.
            If True, enables verbose output.

    Examples
    --------
    >>> import matplotlib.pyplot as plt
    >>> from MDAnalysis import Universe
    >>> from WatAnalysis.workflow import PlanarInterfaceAnalysisBase, DensityAnalysis
    >>>
    >>> # set up universe and surface indices...
    >>>
    >>> obj = PlanarInterfaceAnalysisBase(
    ...     universe=universe,
    ...     surf_ids=surf_ids,
    ...     workflow=[
    ...         DensityAnalysis(selection="name O", mol_mass=15.999, label="oxygen"),
    ...         DensityAnalysis(selection="name H", mol_mass=1.008, label="hydrogen"),
    ...     ],
    ... )
    >>> obj.run()
    >>>
    >>> plt.plot(obj.workflow[0].results.bins, obj.workflow[0].results.density)
    >>> plt.show()
    """

    def __init__(
        self,
        universe: Universe,
        surf_ids: Union[List, np.ndarray],
        axis: int = 2,
        workflow: list = [],
        **kwargs,
    ):
        self.universe = universe
        trajectory = self.universe.trajectory
        super().__init__(trajectory, verbose=kwargs.get("verbose", False))
        self.n_frames = len(trajectory)
        self.surf_ids = surf_ids
        self.axis = axis
        self.workflow = workflow

        self.cross_area = None
        self.r_surf_lo = None
        self.r_surf_hi = None
        self.r_surf_ref = None

        self.data_requirements: Dict[str, DataRequirement] = {}
        for task in self.workflow:
            self.data_requirements.update(task.data_requirements)

    def _prepare(self):
        # Initialize empty arrays
        self.cross_area = 0.0
        self.r_surf_lo = np.zeros(self.n_frames)
        self.r_surf_hi = np.zeros(self.n_frames)

        for k, v in self.data_requirements.items():
            if v.atomic:
                ag = self.universe.select_atoms(v.selection)
                a = np.full((self.n_frames, ag.n_atoms, v.dim), v.value)
            else:
                a = np.full((self.n_frames, v.dim), v.value)
            setattr(self, k, a)

        for task in self.workflow:
            task._prepare(self)

    def _single_frame(self):
        ts_box = self._ts.dimensions
        ts_area = Cell.new(ts_box).area(self.axis)
        self.cross_area += ts_area

        coords = self._ts.positions

        # Absolute surface positions
        r_surf_lo = coords[self.surf_ids[0], self.axis]
        r_surf_hi = coords[self.surf_ids[1], self.axis]
        box_length = ts_box[self.axis]
        # Use MIC in case part of the surface crosses the cell boundaries
        self.r_surf_lo[self._frame_index] = utils.mic_1d(
            r_surf_lo, box_length, ref=r_surf_lo[0]
        ).mean()
        self.r_surf_hi[self._frame_index] = utils.mic_1d(
            r_surf_hi, box_length, ref=r_surf_hi[0]
        ).mean()

        # Update flags for the intermediate arrays
        for v in self.data_requirements.values():
            v.set_update_flag(False)

        for task in self.workflow:
            task._single_frame(self)

    def _conclude(self):
        # Average surface area
        self.cross_area /= self.n_frames

        box_length = self.universe.dimensions[self.axis]
        # Step 1: Set r_surf_lo as the reference point (zero)
        r_surf_lo = np.zeros(self.r_surf_lo.shape)
        # Step 2: Calculate z2 positions relative to z1 using minimum image convention
        # By setting ref=box_length/2, all positions are within one cell length positive of z1
        r_surf_hi = utils.mic_1d(
            self.r_surf_hi - self.r_surf_lo,
            box_length=box_length,
            ref=box_length / 2,
        )
        # Update attributes to the final relative coordinates
        self.r_surf_ref = self.r_surf_lo.copy()
        self.r_surf_lo = r_surf_lo
        self.r_surf_hi = r_surf_hi

        for task in self.workflow:
            task._conclude(self)


class SingleAnalysis(ABC):
    def __init__(
        self,
    ) -> None:
        self.results = Results()
        self.data_requirements: Dict[str, DataRequirement] = {}

    @abstractmethod
    def _prepare(self, analyser: PlanarInterfaceAnalysisBase):
        pass

    @abstractmethod
    def _single_frame(self, analyser: PlanarInterfaceAnalysisBase):
        pass

    @abstractmethod
    def _conclude(self, analyser: PlanarInterfaceAnalysisBase):
        pass

    @staticmethod
    def _setattr(analyser: PlanarInterfaceAnalysisBase, k: str, v: Any):
        if not hasattr(analyser, k):
            setattr(analyser, k, v)


class DataRequirement:
    """
    Class for intermediate data requirements for analysis tasks.
    """

    def __init__(
        self,
        var_name: str,
        atomic: bool,
        dim: int,
        value: Any = 0.0,
        selection: str = "all",
    ) -> None:
        self.var_name = var_name
        self.atomic = atomic
        self.dim = dim
        self.value = value
        self.selection = selection

        # flag for updating the intermediate arrays
        self.update_flag = False

    def set_update_flag(self, flag: bool):
        self.update_flag = flag
