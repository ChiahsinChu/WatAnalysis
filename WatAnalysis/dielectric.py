# SPDX-License-Identifier: LGPL-3.0-or-later
import numpy as np
from MDAnalysis.analysis.base import AnalysisBase
from MDAnalysis.analysis.dielectric import DielectricConstant
from MDAnalysis.exceptions import NoDataError
from MDAnalysis.units import constants, convert

from WatAnalysis.preprocess import make_selection, make_selection_two


class InverseDielectricConstant(AnalysisBase):
    def __init__(
        self,
        universe,
        bins,
        axis="z",
        temperature=330,
        make_whole=True,
        verbose=False,
        **kwargs,
    ) -> None:
        super().__init__(universe.trajectory, verbose)

        self.universe = universe
        self.atoms = universe.atoms
        self.bins = bins
        self.nbins = len(bins) - 1
        axis_dict = {"x": 0, "y": 1, "z": 2}
        self.axis = axis_dict[axis]
        self.temperature = temperature
        self.make_whole = make_whole
        self.kwargs = kwargs

    def _prepare(self):
        if not hasattr(self.atoms, "charges"):
            raise NoDataError("No charges defined given atomgroup.")

        if not np.allclose(
            self.atoms.total_charge(compound="fragments"), 0.0, atol=1e-5
        ):
            raise NotImplementedError(
                "Analysis for non-neutral systems or"
                " systems with free charges are not"
                " available."
            )

        self.results.m = np.zeros((self.nbins))
        self.results.mM = np.zeros((self.nbins))
        self.results.M = 0
        self.results.M2 = 0
        self.volume = 0

        self.ags = []
        for ii in range(self.nbins):
            sel_region = [self.bins[ii], self.bins[ii + 1]]
            select = make_selection(sel_region=sel_region, **self.kwargs)
            self.ags.append(self.universe.select_atoms(select, updating=True))

    def _single_frame(self):
        ave_axis = np.delete(np.arange(3), self.axis)
        ts_area = self._ts.dimensions[ave_axis[0]] * self._ts.dimensions[ave_axis[1]]
        ts_volume = self._ts.volume
        self.volume += ts_volume

        if self.make_whole:
            self.atoms.unwrap()

        M = np.dot(self.atoms.charges, self.atoms.positions)[self.axis]
        self.results.M += M
        self.results.M2 += M * M

        def _single_bin(ii, make_whole, ags, bins, axis, results):
            if make_whole:
                ags[ii].unwrap()
            bin_volume = ts_area * (bins[ii + 1] - bins[ii]) * 2
            m = np.dot(ags[ii].charges, ags[ii].positions)[axis] / bin_volume
            results.m[ii] += m
            results.mM[ii] += m * M
            # print(results)

        for ii in range(self.nbins):
            _single_bin(
                ii, self.make_whole, self.ags, self.bins, self.axis, self.results
            )

    def _conclude(self):
        self.results.m /= self.n_frames
        self.results.mM /= self.n_frames
        self.results.M /= self.n_frames
        self.results.M2 /= self.n_frames
        self.volume /= self.n_frames

        x_fluct = self.results.mM - self.results.m * self.results.M
        M_fluct = self.results.M2 - self.results.M * self.results.M

        self.results.eps = self.results.fluct / (
            convert(constants["Boltzman_constant"], "kJ/mol", "eV")
            * self.temperature
            * self.volume
            * constants["electric_constant"]
        )

        const = (
            convert(constants["Boltzman_constant"], "kJ/mol", "eV")
            * self.temperature
            * constants["electric_constant"]
            / 3.0
        )
        self.results.inveps = 1 - x_fluct / (const + M_fluct / self.volume)


class ParallelInverseDielectricConstant(InverseDielectricConstant):
    def __init__(
        self,
        universe,
        bins,
        axis="z",
        temperature=330,
        make_whole=True,
        verbose=False,
        **kwargs,
    ) -> None:
        super().__init__(
            universe, bins, axis, temperature, make_whole, verbose, **kwargs
        )
        # parallel value initial
        self.para = None
        self._para_region = None

    def _conclude(self):
        pass

    def _parallel_init(self, *args, **kwargs):
        start = self._para_region.start
        stop = self._para_region.stop
        step = self._para_region.step
        self._setup_frames(self._trajectory, start, stop, step)
        self._prepare()

    def run(self, start=None, stop=None, step=None, verbose=None):
        # self._trajectory._reopen()
        if verbose == True:
            print(" ", end="")
        super().run(start, stop, step, verbose)

        if self.para:
            block_result = self._para_block_result()
            if block_result == None:
                raise ValueError(
                    "in parallel, block result has not been defined or no data output!"
                )
            # logger.info("block_anal finished.")
            return block_result

    def _para_block_result(self):
        return [
            self.results.m,
            self.results.mM,
            self.results.M,
            self.results.M2,
            self.volume,
        ]

    def _parallel_conclude(self, rawdata):
        # set attributes for further analysis
        method_attr = rawdata[-1]
        del rawdata[-1]
        self.start = method_attr[0]
        self.stop = method_attr[1]
        self.step = method_attr[2]
        self.frames = np.arange(self.start, self.stop, self.step)
        self.n_frames = len(self.frames)

        self.results["m"] = np.zeros((self.nbins))
        self.results["mM"] = np.zeros((self.nbins))
        self.results["M"] = 0
        self.results["M2"] = 0
        self.volume = 0

        for single_data in rawdata:
            self.results["m"] += single_data[0]
            self.results["mM"] += single_data[1]
            self.results["M"] += single_data[2]
            self.results["M2"] += single_data[3]
            self.volume += single_data[4]

        self.results["m"] /= self.n_frames
        self.results["mM"] /= self.n_frames
        self.results["M"] /= self.n_frames
        self.results["M2"] /= self.n_frames
        self.volume /= self.n_frames

        x_fluct = self.results["mM"] - self.results["m"] * self.results["M"]
        M_fluct = self.results["M2"] - self.results["M"] * self.results["M"]
        const = (
            convert(constants["Boltzman_constant"], "kJ/mol", "eV")
            * self.temperature
            * constants["electric_constant"]
        )
        self.results["inveps"] = 1 - x_fluct / (const + M_fluct / self.volume)

        return "FINISH PARA CONCLUDE"


class DeprecatedDC(DielectricConstant):
    def __init__(
        self, universe, bins, temperature=330, make_whole=True, verbose=False, **kwargs
    ) -> None:
        self.universe = universe
        self.bins = bins
        self.nbins = len(bins) - 1
        self.kwargs = kwargs
        super().__init__(universe.atoms, temperature, make_whole, verbose=verbose)

    def _prepare(self):
        super()._prepare()

        # reset
        self.volume = np.zeros(self.nbins)
        self.results.M = np.zeros((self.nbins, 3))
        self.results.M2 = np.zeros((self.nbins, 3))
        self.results.fluct = np.zeros((self.nbins, 3))
        self.results.eps = np.zeros((self.nbins, 3))
        self.results.eps_mean = np.zeros(self.nbins)

        self.ags = []
        for ii in range(self.nbins):
            sel_region = [self.bins[ii], self.bins[ii + 1]]
            select = make_selection_two(sel_region=sel_region, **self.kwargs)
            self.ags.append(self.universe.select_atoms(select[0], updating=True))
            self.ags.append(self.universe.select_atoms(select[1], updating=True))

    def _single_frame(self):
        for ii in range(self.nbins):
            # lower surface
            if self.make_whole:
                self.ags[2 * ii].unwrap()

            # volume of each bin rather than the whole system
            # self.volume += self.atomgroup.universe.trajectory.ts.volume
            self.volume[ii] += (
                self._ts.dimensions[0]
                * self._ts.dimensions[1]
                * (self.bins[ii + 1] - self.bins[ii])
            )

            M = np.dot(self.ags[2 * ii].charges, self.ags[2 * ii].positions)
            self.results.M[ii] += M
            self.results.M2[ii] += M * M

            # upper surface
            if self.make_whole:
                self.ags[2 * ii + 1].unwrap()

            # volume of each bin rather than the whole system
            # self.volume += self.atomgroup.universe.trajectory.ts.volume
            self.volume[ii] += (
                self._ts.dimensions[0]
                * self._ts.dimensions[1]
                * (self.bins[ii + 1] - self.bins[ii])
            )

            M = np.dot(
                self.ags[2 * ii + 1].charges, self.ags[2 * ii + 1].positions
            ) * np.array([1.0, 1.0, -1.0])
            self.results.M[ii] += M
            self.results.M2[ii] += M * M

    def _conclude(self):
        self.results.M /= self.n_frames
        self.results.M2 /= self.n_frames
        self.volume /= self.n_frames

        self.results.fluct = self.results.M2 - self.results.M * self.results.M

        self.results.eps = self.results.fluct / (
            convert(constants["Boltzman_constant"], "kJ/mol", "eV")
            * self.temperature
            * np.reshape(self.volume, (self.nbins, 1))
            * constants["electric_constant"]
        )

        self.results.eps_mean = self.results.eps.mean(axis=-1)

        self.results.eps += 1
        self.results.eps_mean += 1
