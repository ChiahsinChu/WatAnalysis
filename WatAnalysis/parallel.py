# SPDX-License-Identifier: LGPL-3.0-or-later
import logging
import time
from multiprocessing import cpu_count

# modules for parallel computing
import dask
import dask.multiprocessing

logger = logging.getLogger(__name__)

dask.config.set(scheduler="processes")
n_jobs = cpu_count()


def parallel_run(universe, ana_base):
    """
    based on MDA user guide
    """
    tot_start = time.time()
    print("TASK START at ", tot_start)
    print(tot_start)

    # generate block
    start = time.time()
    print("START: Generate block")
    print(start)

    n_frames = universe.trajectory.n_frames
    n_blocks = n_jobs  #  it can be any realistic value (0 < n_blocks <= n_jobs)
    n_frames_per_block = n_frames // n_blocks
    blocks = [
        range(i * n_frames_per_block, (i + 1) * n_frames_per_block)
        for i in range(n_blocks - 1)
    ]
    blocks.append(range((n_blocks - 1) * n_frames_per_block, n_frames))
    print(blocks)

    print("END: Generate block")
    print(time.time())
    fmt = "\nUsed Time: {:.3f} seconds"
    print(fmt.format(time.time() - start))

    # compute
    start = time.time()
    print("START: Compute")
    print(start)

    jobs = []
    for bs in blocks:
        jobs.append(run_block(bs, ana_base))
    jobs = dask.delayed(jobs)
    results = jobs.compute()

    print("END: Compute")
    print(time.time())
    fmt = "\nUsed Time: {:.3f} seconds"
    print(fmt.format(time.time() - start))

    print("TASK END at ", time.time())
    fmt = "\nTotal Used Time: {:.3f} seconds"
    print(fmt.format(time.time() - tot_start))

    return results


@dask.delayed
def run_block(blockslice, ana_base):
    """
    Args:
        blockslice: TBC
        ana_base: MDA AnalysisBase Class/Child Class
    Return:
        ana_base with results of block
    """
    # universe.transfer_to_memory(start=blockslice.start, stop=blockslice.stop)
    ana_base.run(start=blockslice.start, stop=blockslice.stop, verbose=True)
    return ana_base


# >>>>>>>>>>>>>>> Ke Xiong Code >>>>>>>>>>>>>>> #
def parallel_exec(singlefunc, start, stop, step, n_proc, *args, **kwargs):
    from concurrent.futures import ProcessPoolExecutor
    from functools import partial

    region = range(start, stop, step)
    blocks = slice_split(region, n_proc)

    if singlefunc.__class__.__name__ == "function":
        run_para = partial(
            _parallel_function_formap, singlefunc=singlefunc, *args, **kwargs
        )

    elif singlefunc.__class__.__name__ == "method":
        # type of method we should tell
        if singlefunc.__self__.__class__.__base__.__base__.__name__ == "AnalysisBase":
            # more variables should be added.

            # it seems not work
            singlefunc.__self__.para = True
            singlefunc.__self__._para_region = region

        run_para = partial(
            _parallel_function_formap, singlefunc=singlefunc, *args, **kwargs
        )

    # co future
    with ProcessPoolExecutor(n_proc) as pool_executor:
        raw_result = list(pool_executor.map(run_para, blocks))
    raw_result.append([start, stop, step])

    # raw_result process
    result = para_raw_data_process(singlefunc, raw_result)
    if not result:
        print("no result process match, raw result will be collected.")
        result = raw_result

    try:
        if singlefunc.__self__.__class__.__base__.__base__.__name__ == "AnalysisBase":
            singlefunc.__self__.para = None
    except:
        pass
    # singlefunc.__self__.ag.universe.trajectory._reopen()

    return result


def slice_split(src, num):
    len_src = len(src)
    div = len_src // num
    mod = len_src % num
    sdx = 0
    splist = []
    for p_len in [1] * mod + [0] * (num - mod):
        edx = p_len + div + sdx
        splist.append(src[sdx:edx])
        sdx = edx
    return splist


def _parallel_function_formap(block, singlefunc, *args, **kwargs):
    start = block.start
    stop = block.stop
    step = block.step

    return singlefunc(start=start, stop=stop, step=step, *args, **kwargs)


def para_raw_data_process(func, rawdata, *args):
    result = []
    if func.__class__.__name__ == "method":
        if func.__name__ == "trans2ase":
            result = []
            for item in rawdata:
                result += item
        elif func.__name__ == "get_memory_slice":
            from MDAnalysis.coordinates.memory import MemoryReader

            coords = []
            for item in rawdata:
                single_coords = item.get_array()
                coords.append(single_coords)
            coords = np.concatenate(coords)
            memory_slice = MemoryReader(coords)
            result = memory_slice
        elif func.__name__ == "get_distance":
            result = np.concatenate(rawdata)
        ##suit for analysis method.
        elif func.__self__.__class__.__base__.__base__.__name__ == "AnalysisBase":
            result = func.__self__._parallel_conclude(rawdata)
        elif func.__self__.__class__.__name__ == "OffsetGenerator":
            result = rawdata

    elif func.__class__.__name__ == "function":
        if func.__name__ == "offset_para_func":
            import numpy as np

            offsets = np.concatenate(rawdata)
            offsets = np.unique(offsets)
            n_frames = len(offsets)
            result = (n_frames, offsets)
        if func.__name__ == "trans2ase":
            result = []
            for item in rawdata:
                result += item
    return result


# <<<<<<<<<<<<<<< Ke Xiong Code <<<<<<<<<<<<<<< #
