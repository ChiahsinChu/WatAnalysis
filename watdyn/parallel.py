import time, logging
# modules for parallel computing
import dask
import dask.multiprocessing
from multiprocessing import cpu_count

logger = logging.getLogger(__name__)

dask.config.set(scheduler='processes')
n_jobs = cpu_count()


@dask.delayed
def run_block(blockslice, ana_base):
    """
    Args:
        blockslice: TBC
        ana_base: MDA AnalysisBase Class/Child Class
    Return:
        ana_base with results of block
    """
    #universe.transfer_to_memory(start=blockslice.start, stop=blockslice.stop)
    ana_base.run(start=blockslice.start, stop=blockslice.stop, verbose=True)
    return ana_base


def parallel_run(universe, ana_base):
    tot_start = time.process_time()
    print("TASK START at ", tot_start)
    print(tot_start)

    # generate block
    start = time.process_time()
    print("START: Generate block")
    print(start)

    n_frames = universe.trajectory.n_frames
    n_blocks = n_jobs  #  it can be any realistic value (0 < n_blocks <= n_jobs)
    n_frames_per_block = n_frames // n_blocks
    blocks = [
        range(i * n_frames_per_block, (i + 1) * n_frames_per_block)
        for i in range(n_blocks - 1)
    ]
    blocks.append(range((n_blocks - 1) * n_frames_per_block, n_frames))
    print(blocks)

    print("END: Generate block")
    print(time.process_time())
    fmt = "\nUsed Time: {:.3f} seconds"
    print(fmt.format(time.process_time() - start))

    # compute
    start = time.process_time()
    print("START: Compute")
    print(start)

    jobs = []
    for bs in blocks:
        jobs.append(run_block(bs, ana_base))
    jobs = dask.delayed(jobs)
    results = jobs.compute()

    print("END: Compute")
    print(time.process_time())
    fmt = "\nUsed Time: {:.3f} seconds"
    print(fmt.format(time.process_time() - start))

    print("TASK END at ", time.process_time())
    fmt = "\nTotal Used Time: {:.3f} seconds"
    print(fmt.format(time.process_time() - tot_start))

    return results
