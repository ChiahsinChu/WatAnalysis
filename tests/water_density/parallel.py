import numpy as np
import time
import MDAnalysis as mda
from MDAnalysis import transformations as trans

from WatAnalysis.base import InterfaceWatDensity
from WatAnalysis.parallel import parallel_exec

# load trajectory
u = mda.Universe("../input_data/interface.psf", "../input_data/trajectory.xyz")
dim = [16.869, 16.869, 41.478, 90, 90, 120]
transform = trans.boxdimensions.set_dimensions(dim)
u.trajectory.add_transformations(transform)

start = time.time()
density = InterfaceWatDensity(universe=u, slab_sel='name Pt', surf_natoms=36)
parallel_exec(density.run, 0, 10000, 1, 4)
fmt = "\nWork Completed! Used Time: {:.3f} seconds"
print(fmt.format(time.time() - start))

output = np.concatenate(([density.results['grids']], [density.results['density']]), axis=0)
np.save("./parallel.npy", output)