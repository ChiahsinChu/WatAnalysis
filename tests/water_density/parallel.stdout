
Work Completed! Used Time: 11.241 seconds
